File: library.py
import json
from book import Book
from member import Member
from typing import Dict, List, Optional

class Library:
    def __init__(self):
        self.books: Dict[str, Book] = {}
        self.members: Dict[str, Member] = {}
        self.next_member_ID: int = 1

    def add_books(self, name: str, author: str, ISBN: str, quantity: int = 1) -> None:
        if ISBN in self.books:
            self.books[ISBN].quantity += quantity
            self.books[ISBN].available_copies += quantity
            print(f"Added {quantity} of books of {name}. Total books of {name} = {self.books[ISBN].quantity}.")
        else:
            new_book = Book(name, author, ISBN, quantity)
            self.books[ISBN] = new_book
            print(f"{name} titled book written by {author} is added to the library.")

    def remove_book(self, ISBN: str) -> None:
        if ISBN in self.books:
            book_title = self.books[ISBN].title
            del self.books[ISBN]
            print(f"Book with {ISBN} isbn number and titled {book_title} removed from the library.")
        else:
            print(f"Book with ISBN {ISBN} not found.")

    def register_member(self, name: str) -> Member:
        member_id = f"M{self.next_member_ID:04d}"
        new_member = Member(name, member_id)
        self.members[member_id] = new_member
        self.next_member_ID += 1
        return new_member
    
    def deregister_member(self, member_ID: str) -> None:
        if member_ID in self.members:
            member = self.members[member_ID]
            if not member.borrowed_books:
                del self.members[member_ID]
                print(f"Member {member.name} with member ID {member_ID} deregistered.")
            else:
                print(f"Member {member.name} with member ID {member_ID} cannot be deregistered. They still have some books borrowed.")
        else:
            print(f"Member with ID {member_ID} not found.")

    def borrow_book(self, member_ID: str, ISBN: str) -> None:
        member = self.members.get(member_ID)
        book = self.books.get(ISBN)

        if not member:
            print(f"Error: Member with ID {member_ID} not found.")
            return
        if not book:
            print(f"Error: Book with ISBN {ISBN} not found.")
            return

        if book.borrow_copy():
            if member.borrow_book(book):
                print(f"{member.name} has successfully borrowed {book.title}.")
            else:
                book.return_copy()
                print(f"{member.name} already has a copy of {book.title}.")
        else:
            print(f"{book.title} is currently not available for borrowing.")

    def return_book(self, member_ID: str, ISBN: str) -> None:
        member = self.members.get(member_ID)
        book = self.books.get(ISBN)

        if not member:
            print(f"Error: Member with ID {member_ID} not found.")
            return
        if not book:
            print(f"Error: Book with ISBN {ISBN} not found.")
            return
        if member.return_book(book):
            book.return_copy()
            print(f"{member.name} has successfully returned {book.title}.")
        else:
            print(f"{member.name} did not borrow {book.title}.")

    def search_book(self, query: str, search_by: str = "name") -> List[Book]:
        results = []
        query_lower = query.lower()
        for ISBN, book in self.books.items():
            if search_by == "name" and query_lower in book.title.lower():
                results.append(book)
            elif search_by == "author" and query_lower in book.author.lower():
                results.append(book)
            elif search_by == "ISBN" and query_lower == book.ISBN.lower():
                results.append(book)
        return results
    
    def view_member_borrowed_books(self, member_ID: str) -> None:
        member = self.members.get(member_ID)
        if member:
            if member.borrowed_books:
                print(f"\nBooks borrowed by {member.name} (ID: {member_ID}):")
                for book in member.borrowed_books:
                    print(f"  - {book.title} by {book.author} (ISBN: {book.ISBN})")
            else:
                print(f"{member.name} (ID: {member_ID}) has no books currently borrowed.")
        else:
            print(f"Member with ID '{member_ID}' not found.")

    def display_all_books(self) -> None:
        if not self.books:
            print("\nNo books in the library.")
            return
        print("\n--- All Books in Library ---")
        for ISBN, book in self.books.items():
            print(book)
        print("----------------------------")

    def display_all_members(self) -> None:
        if not self.members:
            print("\nNo members registered.")
            return
        print("\n--- All Registered Members ---")
        for member_ID, member in self.members.items():
            print(member)
        print("------------------------------")

    def save_data(self, filename: str = "library_data.json") -> None:
        data = {
            "books": {ISBN: book.to_dict() for ISBN, book in self.books.items()},
            "members": {member_ID: member.to_dict() for member_ID, member in self.members.items()},
            "next_member_ID": self.next_member_ID
        }
        with open(filename, 'w') as f:
            json.dump(data, f, indent=4)
        print("Library data saved successfully.")

    def load_data(self, filename: str = "library_data.json") -> None:
        try:
            with open(filename, 'r') as f:
                data = json.load(f)

            self.books = {ISBN: Book.from_dict(book_data) for ISBN, book_data in data.get("books", {}).items()}

            self.members = {}
            for member_ID, member_data in data.get("members", {}).items():
                member = Member.from_dict(member_data, self.books)
                self.members[member_ID] = member
                for ISBN in member_data.get('borrowed_book_isbns', []):
                    if ISBN in self.books:
                        member.borrowed_books.append(self.books[ISBN])
                    else:
                        print(f"Warning: Book with ISBN {ISBN} borrowed by {member.name} not found in library inventory.")

            self.next_member_ID = data.get("next_member_ID", 1)
            print("Library data loaded successfully.")
        except FileNotFoundError:
            print("No saved library data found. Starting with an empty library.")
        except Exception as e:
            print(f"Error loading data: {e}")

File: js/library.js
/**
 * Library Class - Enhanced central management system
 */
class Library {
    constructor() {
        this.books = {}; // ISBN -> Book object
        this.members = {}; // Member ID -> Member object
        this.nextMemberID = 1;
        this.reservations = []; // Book reservations
        this.categories = ['General', 'Fiction', 'Non-Fiction', 'Science', 'Technology', 'History', 'Biography', 'Fantasy', 'Mystery', 'Romance', 'Self-Help', 'Children'];
        this.libraryName = 'Digital Library Management System';
        this.settings = {
            borrowPeriodDays: 14,
            maxBooksPerMember: 5,
            finePerDay: 0.50,
            allowReservations: true
        };
    }

    /**
     * Generate next member ID (M0001, M0002, etc.)
     */
    generateMemberID() {
        const id = `M${String(this.nextMemberID).padStart(4, '0')}`;
        this.nextMemberID++;
        return id;
    }

    /**
     * Add a new book or update quantity if exists
     */
    addBook(name, author, ISBN, quantity, category = 'General', description = '', coverImage = null) {
        if (this.books[ISBN]) {
            // Book exists, update quantity
            this.books[ISBN].quantity += parseInt(quantity);
            this.books[ISBN].available += parseInt(quantity);
            return { success: true, message: `Updated quantity for "${name}". New total: ${this.books[ISBN].quantity}`, isNew: false };
        } else {
            // New book
            this.books[ISBN] = new Book(name, author, ISBN, quantity, category, description, coverImage);
            return { success: true, message: `Added new book: "${name}"`, isNew: true };
        }
    }

    /**
     * Remove a book by ISBN
     */
    removeBook(ISBN) {
        if (!this.books[ISBN]) {
            return { success: false, message: 'Book not found' };
        }

        const bookName = this.books[ISBN].name;
        delete this.books[ISBN];

        // Remove any reservations for this book
        this.reservations = this.reservations.filter(r => r.isbn !== ISBN);

        return { success: true, message: `Removed book: "${bookName}"` };
    }

    /**
     * Register a new member
     */
    registerMember(name, phone, email, address, profilePicture = null) {
        const id = this.generateMemberID();
        this.members[id] = new Member(name, phone, email, address, id, profilePicture);
        return { success: true, message: `Member registered successfully! ID: ${id}`, memberId: id };
    }

    /**
     * Deregister a member
     */
    deregisterMember(memberID) {
        if (!this.members[memberID]) {
            return { success: false, message: 'Member not found' };
        }

        const member = this.members[memberID];
        if (member.getCurrentBorrowedBooks().length > 0) {
            return { success: false, message: 'Cannot deregister member with borrowed books. Please return all books first.' };
        }

        if (member.currentFines > 0) {
            return { success: false, message: `Cannot deregister member with outstanding fines: $${member.currentFines.toFixed(2)}` };
        }

        const memberName = member.name;
        delete this.members[memberID];

        // Remove any reservations by this member
        this.reservations = this.reservations.filter(r => r.memberId !== memberID);

        return { success: true, message: `Deregistered member: ${memberName}` };
    }

    /**
     * Borrow a book
     */
    borrowBook(memberID, ISBN) {
        // Validate member
        if (!this.members[memberID]) {
            return { success: false, message: 'Member not found' };
        }

        // Validate book
        if (!this.books[ISBN]) {
            return { success: false, message: 'Book not found' };
        }

        const member = this.members[memberID];
        const book = this.books[ISBN];

        // Check max books limit
        if (member.getCurrentBorrowedBooks().length >= this.settings.maxBooksPerMember) {
            return { success: false, message: `Maximum ${this.settings.maxBooksPerMember} books allowed per member` };
        }

        // Check if book is available
        if (!book.isAvailable()) {
            return { success: false, message: 'Book is not available' };
        }

        // Check if member already has this book
        if (member.getBorrowedBookISBNs().includes(ISBN)) {
            return { success: false, message: 'Member has already borrowed this book' };
        }

        // Calculate due date
        const dueDate = new Date();
        dueDate.setDate(dueDate.getDate() + this.settings.borrowPeriodDays);

        // Perform borrow
        if (book.borrowBook() && member.borrowBook(ISBN, book.name, dueDate)) {
            // Remove reservation if exists
            this.reservations = this.reservations.filter(r => !(r.isbn === ISBN && r.memberId === memberID));

            return {
                success: true,
                message: `"${book.name}" borrowed successfully!`,
                dueDate: dueDate.toLocaleDateString()
            };
        }

        return { success: false, message: 'Failed to borrow book' };
    }

    /**
     * Return a book
     */
    returnBook(memberID, ISBN) {
        // Validate member
        if (!this.members[memberID]) {
            return { success: false, message: 'Member not found' };
        }

        // Validate book
        if (!this.books[ISBN]) {
            return { success: false, message: 'Book not found' };
        }

        const member = this.members[memberID];
        const book = this.books[ISBN];

        // Check if member has borrowed this book
        if (!member.getBorrowedBookISBNs().includes(ISBN)) {
            return { success: false, message: 'Member has not borrowed this book' };
        }

        // Perform return
        if (member.returnBook(ISBN) && book.returnBook()) {
            const borrowRecord = member.booksBorrowed.find(b => b.isbn === ISBN && b.returnDate);

            let message = `"${book.name}" returned successfully!`;
            if (borrowRecord && borrowRecord.isOverdue && borrowRecord.fine) {
                message += ` Late fee: $${borrowRecord.fine.toFixed(2)}`;
            }

            // Notify if anyone has reserved this book
            const reservation = this.reservations.find(r => r.isbn === ISBN);
            if (reservation) {
                message += ` Note: Book is reserved by ${this.members[reservation.memberId].name}`;
            }

            return { success: true, message: message };
        }

        return { success: false, message: 'Failed to return book' };
    }

    /**
     * Reserve a book
     */
    reserveBook(memberID, ISBN) {
        if (!this.settings.allowReservations) {
            return { success: false, message: 'Reservations are not allowed' };
        }

        if (!this.members[memberID]) {
            return { success: false, message: 'Member not found' };
        }

        if (!this.books[ISBN]) {
            return { success: false, message: 'Book not found' };
        }

        // Check if already reserved by this member
        if (this.reservations.some(r => r.isbn === ISBN && r.memberId === memberID)) {
            return { success: false, message: 'You have already reserved this book' };
        }

        this.reservations.push({
            isbn: ISBN,
            memberId: memberID,
            reservationDate: new Date().toISOString()
        });

        return { success: true, message: `Book "${this.books[ISBN].name}" reserved successfully!` };
    }

    /**
     * Cancel reservation
     */
    cancelReservation(memberID, ISBN) {
        const initialLength = this.reservations.length;
        this.reservations = this.reservations.filter(r => !(r.isbn === ISBN && r.memberId === memberID));

        if (this.reservations.length < initialLength) {
            return { success: true, message: 'Reservation cancelled' };
        }

        return { success: false, message: 'Reservation not found' };
    }

    /**
     * Search books by title, author, ISBN, or category
     */
    searchBooks(query, searchBy = 'all') {
        query = query.toLowerCase().trim();
        const results = [];

        for (let ISBN in this.books) {
            const book = this.books[ISBN];

            if (searchBy === 'all') {
                if (book.name.toLowerCase().includes(query) ||
                    book.author.toLowerCase().includes(query) ||
                    book.ISBN.toLowerCase().includes(query) ||
                    book.category.toLowerCase().includes(query)) {
                    results.push(book);
                }
            } else if (searchBy === 'title' && book.name.toLowerCase().includes(query)) {
                results.push(book);
            } else if (searchBy === 'author' && book.author.toLowerCase().includes(query)) {
                results.push(book);
            } else if (searchBy === 'isbn' && book.ISBN.toLowerCase().includes(query)) {
                results.push(book);
            } else if (searchBy === 'category' && book.category.toLowerCase().includes(query)) {
                results.push(book);
            }
        }

        return results;
    }

    /**
     * Get books by category
     */
    getBooksByCategory(category) {
        return Object.values(this.books).filter(book => book.category === category);
    }

    /**
     * Get all books as array
     */
    getAllBooks() {
        return Object.values(this.books);
    }

    /**
     * Get all members as array
     */
    getAllMembers() {
        return Object.values(this.members);
    }

    /**
     * Get member's borrowed books with details
     */
    getMemberBorrowedBooks(memberID) {
        if (!this.members[memberID]) {
            return null;
        }

        const member = this.members[memberID];
        const borrowedBooks = member.getCurrentBorrowedBooks();

        return borrowedBooks.map(b => {
            const book = this.books[b.isbn];
            return {
                ...b,
                bookDetails: book ? {
                    name: book.name,
                    author: book.author,
                    category: book.category,
                    coverImage: book.coverImage
                } : null
            };
        });
    }

    /**
     * Get library statistics
     */
    getStatistics() {
        const books = this.getAllBooks();
        const members = this.getAllMembers();

        let totalBooks = 0;
        let availableBooks = 0;
        let borrowedBooks = 0;

        books.forEach(book => {
            totalBooks += book.quantity;
            availableBooks += book.available;
            borrowedBooks += (book.quantity - book.available);
        });

        let totalMembers = members.length;
        let activeMembers = members.filter(m => m.getCurrentBorrowedBooks().length > 0).length;
        let overdueBooks = 0;
        let totalFines = 0;

        members.forEach(member => {
            member.updateOverdueStatus();
            overdueBooks += member.getOverdueBooks().length;
            totalFines += member.currentFines;
        });

        return {
            totalBooks: totalBooks,
            uniqueTitles: books.length,
            availableBooks: availableBooks,
            borrowedBooks: borrowedBooks,
            totalMembers: totalMembers,
            activeMembers: activeMembers,
            overdueBooks: overdueBooks,
            totalFines: totalFines.toFixed(2),
            totalReservations: this.reservations.length,
            categories: this.categories.length
        };
    }

    /**
     * Get popular books (most borrowed)
     */
    getPopularBooks(limit = 5) {
        return this.getAllBooks()
            .sort((a, b) => b.totalBorrowCount - a.totalBorrowCount)
            .slice(0, limit);
    }

    /**
     * Get top rated books
     */
    getTopRatedBooks(limit = 5) {
        return this.getAllBooks()
            .filter(book => book.ratings.length > 0)
            .sort((a, b) => b.getAverageRating() - a.getAverageRating())
            .slice(0, limit);
    }

    /**
     * Convert to plain object for storage
     */
    toJSON() {
        return {
            books: this.books,
            members: this.members,
            nextMemberID: this.nextMemberID,
            reservations: this.reservations,
            categories: this.categories,
            libraryName: this.libraryName,
            settings: this.settings
        };
    }

    /**
     * Load library from plain object
     */
    static fromJSON(data) {
        const library = new Library();

        // Load books
        if (data.books) {
            for (let isbn in data.books) {
                library.books[isbn] = Book.fromJSON(data.books[isbn]);
            }
        }

        // Load members
        if (data.members) {
            for (let id in data.members) {
                library.members[id] = Member.fromJSON(data.members[id]);
            }
        }

        library.nextMemberID = data.nextMemberID || 1;
        library.reservations = data.reservations || [];
        library.categories = data.categories || library.categories;
        library.libraryName = data.libraryName || library.libraryName;
        library.settings = data.settings || library.settings;

        return library;
    }
}

// Export for use in other modules
if (typeof module !== 'undefined' && module.exports) {
    module.exports = Library;
}
